File: java2/src/mal/reader.java
package mal;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

class reader {

  static mal read_str(String val) {
    return read_form(new Reader(tokenizer(val)));
  }

  static mal read_form(Reader reader) {
    String token = reader.peek();
    if (token == null) {
      throw new EOFException();
    }
    mal form;
    switch (token) {
      case "'":
      reader.next();
      return read_shorthand(new symbol("quote"), read_form(reader));
      case "`":
      reader.next();
      return read_shorthand(new symbol("quasiquote"), read_form(reader));
      case "~":
      reader.next();
      return read_shorthand(new symbol("unquote"), read_form(reader));
      case "@":
      reader.next();
      return read_shorthand(new symbol("deref"), read_form(reader));
      case "~@":
      reader.next();
      return read_shorthand(new symbol("splice-unquote"), read_form(reader));
      case "^":
      reader.next();
      mal meta = read_form(reader);
      mal data = read_form(reader);
      return read_with_meta(new symbol("with-meta"), meta, data);
      case "(":
      case "[":
      case "{":
        form = read_list(reader);
        break;
      default:
        form = read_atom(reader);
    }
    return form;
  }

  private static list read_with_meta(symbol symbol, mal meta, mal data) {
    list list = new list();
    list.add(symbol);
    list.add(data);
    list.add(meta);
    return list;
  }

  static list read_shorthand(symbol symbol, mal data) {
    list list = new list();
    list.add(symbol);
    list.add(data);
    return list;
  }

  static list read_list(Reader reader) {
    String token = reader.next();
    switch (token) {
      case "[":
        vector vector = new vector();

        while (!reader.peek().equals("]")) {
          vector.add(read_form(reader));

          if (reader.peek() == null) {
            throw new EOFException("]");
          }
        }
        reader.next();
        return vector;
      case "{":
        List<mal> data = new ArrayList<mal>();
        while (!reader.peek().equals("}")) {
          data.add(read_form(reader));

          if (reader.peek() == null) {
            throw new EOFException("}");
          }
        }
        reader.next();

        return new hash_map(data);
      default:
        list list = new list();
        while (!reader.peek().equals(")")) {
          list.add(read_form(reader));

          if (reader.peek() == null) {
            throw new EOFException(")");
          }
        }
        reader.next();
        return list;
    }
  }

  static mal read_atom(Reader reader) {
    String token = reader.next();

    try {
      return new number(Long.parseLong(token));
    } catch (NumberFormatException e) {
      // do nothing
    }

    if (token.startsWith("\"")) {
      String content = token.substring(1, token.length() - 1);

      Pattern pattern = Pattern.compile("\\\\(.)");
      Matcher matcher = pattern.matcher(content);

      StringBuffer sb = new StringBuffer();
      while (matcher.find()) {
        if (matcher.group(1).equals("n")) {
          matcher.appendReplacement(sb, "\n");
        }else if (matcher.group(1).equals("\\")) {
          matcher.appendReplacement(sb, "\\\\");
        }else {
          matcher.appendReplacement(sb, matcher.group(1));
        }
      }
      matcher.appendTail(sb);
      return new str(sb.toString());
    }

    if (token.startsWith(":")) {
      return new keyword(token);
    }

    switch (token) {
      case "true":
        return new True();
      case "false":
        return new False();
      case "nil":
        return new nil();
      default:
        return new symbol(token);
    }
  }

  static List<String> tokenizer(String val) {
    List<String> tokens = new ArrayList<>();
    Pattern pattern = Pattern.compile("[\\s,]*(~@|[\\[\\]{}()'`~^@]|\"(?:\\\\.|[^\\\\\"])*\"|;.*|[^\\s\\[\\]{}('\"`,;)]*)");
    Matcher matcher = pattern.matcher(val);

    while (matcher.find()) {
      String token = matcher.group(1);
      if (token != null &&
          !token.equals("") &&
          !(token.charAt(0) == ';')) {
        tokens.add(token);
      }
    }

    return tokens;
  }

  static class Reader {
    List<String> tokens;
    int position;

    Reader(List<String> tokens) {
      this.tokens = tokens;
      this.position = 0;
    }

    String next() {
      if (position == tokens.size()) {
        return null;
      }
      String token = tokens.get(position);
      position++;
      return token;
    }

    String peek() {
      if (position == tokens.size()) {
        return null;
      }
      return tokens.get(position);
    }
  }

  static class EOFException extends RuntimeException {
    final String msg;

    EOFException() {
      this("");
    }

    EOFException(String msg) {
      this.msg = msg;
    }
  }
}

File: impls/cs.2/reader.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

namespace mal
{
    class Reader
    {
        static string TOKENIZER_REGEX = "[\\s,]*(~@|[\\[\\]{}()'`~^@]|\"(?:\\\\.|[^\\\\\"])*\"?|;.*|[^\\s\\[\\]{}('\"`,;)]*)";
        static Regex TOKENIZER_INSTANCE = new Regex(TOKENIZER_REGEX);
        static Dictionary<char, string> UNESCAPE = new Dictionary<char, string>()
        {
            {'n', "\n"}, {'"', "\""}, {'\\', "\\"}, {'t', "\t"}
        };
        static Dictionary<string, string> QUOTING = new Dictionary<string, string>()
        {
            {"'", "quote"}, {"`", "quasiquote"}, {"~", "unquote"}, {"~@", "splice-unquote"}, {"@", "deref"}
        };
        static Dictionary<string, MalType> LITERALS = new Dictionary<string, MalType>()
        {
            {"nil", MalNil.MAL_NIL}, {"true", MalBoolean.MAL_TRUE}, {"false", MalBoolean.MAL_FALSE}
        };

        IList<string> tokens;
        int position;

        Reader(IList<string> tokens)
        {
            this.tokens = tokens;
            this.position = 0;
        }

        string next()
        {
            if (position >= tokens.Count)
            {
                return null;
            }
            else
            {
                string val = peek();
                position++;
                return val;
            }
        }

        string peek()
        {
            return (position < tokens.Count) ? tokens[position] : null;
        }

        static IList<string> tokenize(string input)
        {
            IList<string> tokens = new List<string>();
            Match match = TOKENIZER_INSTANCE.Match(input);
            while (match.Success)
            {
                string value = match.Groups[1].Value;
                if (value?.Length > 0)
                {
                    tokens.Add(value);
                }
                match = match.NextMatch();
            }

            return tokens;
        }

        public static MalType read_str(string input)
        {
            var tokens = tokenize(input);
            var reader = new Reader(tokens);
            return read_form(reader);
        }

        public static MalType read_form(Reader reader)
        {
            string first = reader.peek();
            if (first == null) return null; // signal that we don't want to print back
            if (first.StartsWith(";")){
                // drop the token if it's a comment, continue with the next token
                reader.next();
                return read_form(reader);
            }
            if (first == "^") // expect two other forms
            {
                reader.next(); // drop the '^'
                MalHashmap metadata = read_hashmap(reader);
                MalType value = read_form(reader);
                List<MalType> items = new List<MalType>() { new MalSymbol("with-meta"), value, metadata };
                MalList listWithMeta = new MalList(items);
                return listWithMeta;
            }
            else if (QUOTING.ContainsKey(first))
            {
                string quotation = QUOTING.GetValueOrDefault(reader.next(), null);
                MalType quoted = read_form(reader);
                List<MalType> items = new List<MalType>()
                {
                    new MalSymbol(quotation),
                    quoted
                };
                MalList quotedList = new MalList(items);
                return quotedList;
            }
            else if (CLOSING_BRACKETS.ContainsValue(first))
            {
                return null; // empty form
            }
            else if (first == "{")
            {
                return read_hashmap(reader);
            }
            else if (first == "[")
            {
                return read_vector(reader);
            }
            else if (CLOSING_BRACKETS.ContainsKey(first))
            {
                return read_list(reader);
            }
            else
            {
                return read_atom(reader);
            }
        }

        public static Dictionary<String, String> CLOSING_BRACKETS = new Dictionary<string, string>() {
            {"(", ")"}, {"[", "]"}, {"{", "}"}
        };
        static MalList read_list(Reader reader)
        {
            string openingBracket = reader.next();
            string closingBracket = CLOSING_BRACKETS.GetValueOrDefault(openingBracket);

            List<MalType> items = new List<MalType>();
            string peeked = null;
            do
            {
                MalType item = read_form(reader);
                if (item != null)
                {
                    items.Add(item);
                }
                peeked = reader.peek();
            } while (peeked != null && peeked != closingBracket);

            // confirm the list has valid closing parens
            var closing = reader.next();
            if (closing == closingBracket)
            {
                return new MalList(items);
            }
            else
            {
                throw new MalException(new MalString("Expression has unbalanced parenthesis"));
            }
        }

        static MalVector read_vector(Reader reader)
        {
            MalList values = read_list(reader);
            return new MalVector(values.items);
        }

        static MalHashmap read_hashmap(Reader reader)
        {
            MalList kvs = read_list(reader);
            if (kvs.items.Count % 2 == 1)
            {
                throw new MalException(new MalString("Hashmap needs an even number of forms"));
            }
            Dictionary<MalType, MalType> pairs = new Dictionary<MalType, MalType>();
            for (int i = 0; i < kvs.items.Count; i += 2)
            {
                MalType key = kvs.items[i];
                MalType val = kvs.items[i + 1];
                pairs.Add(key, val);
            }
            return new MalHashmap(pairs);
        }
        static MalType read_atom(Reader reader)
        {
            // Attempt to parse a number first
            string item = reader.next();
            if (item.StartsWith(":"))
            {
                return new MalKeyword(item.Substring(1));
            }
            else if (LITERALS.ContainsKey(item))
            {
                return LITERALS.GetValueOrDefault(item);
            }
            else if (item.StartsWith("\""))
            {
                if (item.Length > 1 && item.EndsWith("\""))
                {
                    string source = item.Substring(1, item.Length - 2);
                    string unescaped = unescape(source);
                    return new MalString(unescaped); // length minus the quotes
                }
                else
                {
                    throw new MalException(new MalString("String contains unbalanced quotes"));
                }
            }
            try
            {
                int intValue = int.Parse(item);
                return new MalInteger(intValue);
            }
            catch
            {
                return new MalSymbol(item);
            }
        }

        static string unescape(string input)
        {
            bool escaping = false;
            string output = "";
            foreach (char c in input)
            {
                if (!escaping && c == '\\')
                {
                    escaping = true;
                    continue;
                }
                else if (escaping)
                {
                    if (UNESCAPE.ContainsKey(c))
                    {
                        output += UNESCAPE[c];
                        escaping = false;
                    }
                    else
                    {
                        throw new MalException(new MalString("String contains unbalanced escaped characters"));
                    }
                }
                else
                {
                    output += c;
                }

            }
            if (escaping)
            {
                throw new MalException(new MalString("String contains unbalanced escaped characters"));
            }
            return output;
        }
    }
}
